### PR TITLE
feat: Thread time_of_{first,last}_write through Parquet metadata

### PR DESCRIPTION
**Pull request by carols10cents**
This adds tracking of first/last write times of Parquet chunks and persists the data in the Parquet metadata.

### CODE REVIEW

File: generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto [hunk 1]
```diff
@@ -31,6 +31,12 @@ message IoxMetadata {
 
   // Database checkpoint created at the time of the write.
   DatabaseCheckpoint database_checkpoint = 7;
+
+  // Timestamp when this file was first written.
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto:35]: I think this timestamp represents not when the file was written, but the first time the data contained within it was received by IOx. Something like

```suggestion
  // Wallclock timestamp of when the first data in this file was received by IOx.
```

File: generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto [hunk 2]
```diff
@@ -31,6 +31,12 @@ message IoxMetadata {
...
+  FixedSizeTimestamp time_of_first_write = 8;
+
+  // Timestamp when this file was last written.
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto:38]:
```suggestion
  // Wallclock timestamp of when the last data in this file was received by IOx.
```

File: parquet_file/src/metadata.rs [hunk 3]
```diff
@@ -255,6 +252,9 @@ pub struct IoxMetadata {
     /// Timestamp when this file was created.
     pub creation_timestamp: DateTime<Utc>,
 
+    pub time_of_first_write: DateTime<Utc>, // TODO: METADATA_VERSION?
```
Comment (alamb) [inline parquet_file/src/metadata.rs:255]: Is this TODO still relevant? Is it something you plan to address in this PR?

File: parquet_file/src/metadata.rs [hunk 4]
```diff
@@ -505,24 +514,19 @@ impl IoxParquetMetaData {
     }
 
     /// Read IOx statistics (including timestamp range) from parquet metadata.
-    pub fn read_statistics(&self, schema: &Schema, table_name: &str) -> Result<TableSummary> {
-        let mut table_summary_agg: Option<TableSummary> = None;
+    pub fn read_statistics(&self, schema: &Schema) -> Result<Vec<ColumnSummary>> {
```
Comment (alamb) [inline parquet_file/src/metadata.rs:517]: Not needing `table_name` passed in is a nice cleanup

File: server/src/db/chunk.rs [hunk 5]
```diff
@@ -207,15 +207,15 @@ impl DbChunk {
         match &self.state {
             State::MutableBuffer { chunk } => chunk.table_summary().time_of_first_write,
             State::ReadBuffer { chunk, .. } => chunk.table_summary().time_of_first_write,
-            _ => unimplemented!(),
```
Comment (alamb) [inline server/src/db/chunk.rs:210]: 🎉

**Review by alamb — APPROVED** [review]
Looks great to me @carols10cents  -- it might be worth @crepererum reviewing as well as he is likely the most familiar with the `parquet_file` codebase

File: parquet_file/src/metadata.rs (hunk 3 above)
Comment (crepererum) [reply parquet_file/src/metadata.rs:255]: yes, please bump both the metadata version in this module and the catalog transaction version (in the top section of `catalog.rs`).

File: parquet_file/src/metadata.rs [hunk 6]
```diff
@@ -288,13 +288,14 @@ impl IoxMetadata {
...
         // extract creation timestamp
-        let creation_timestamp: DateTime<Utc> = decode_timestamp(
-            proto_msg
-                .creation_timestamp
-                .context(IoxMetadataFieldMissing {
-                    field: "creation_timestamp",
-                })?,
-        )?;
+        let creation_timestamp =
+            decode_timestamp_from_field(proto_msg.creation_timestamp, "creation_timestamp")?;
```
Comment (crepererum) [inline parquet_file/src/metadata.rs:292]: that's a nice API improvement :+1:

**Review by crepererum — APPROVED** [review]
I see you've walked all the way through the parquet metadata stack and the resulting implementation is function-wise exactly what I had imagined and API-wise even improves the current design. Apart from the version bump (see Andrews comment) there's nothing else left for me to say. Good job! :)

File: generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto (hunk 1 above)
Comment (carols10cents) [reply generated_types/protos/influxdata/iox/catalog/v1/parquet_metadata.proto:35]: You're totally right.

File: parquet_file/src/metadata.rs (hunk 3 above)
Comment (carols10cents) [reply parquet_file/src/metadata.rs:255]: Ah yes, I meant to call that out and then forgot. Will do in just a second before merging this!

**Pull request merged**